File: simplified-tests/src/test/java/org/nypl/simplified/tests/opds/OPDSJSONSerializerTest.java
package org.nypl.simplified.tests.opds;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.nypl.simplified.links.Link;
import org.nypl.simplified.opds.core.OPDSAcquisition;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeed;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntryParser;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntryParserType;
import org.nypl.simplified.opds.core.OPDSFeedParser;
import org.nypl.simplified.opds.core.OPDSFeedParserType;
import org.nypl.simplified.opds.core.OPDSJSONParser;
import org.nypl.simplified.opds.core.OPDSJSONParserType;
import org.nypl.simplified.opds.core.OPDSJSONSerializer;
import org.nypl.simplified.opds.core.OPDSJSONSerializerType;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.net.URI;
import java.net.URL;
import java.util.List;

public final class OPDSJSONSerializerTest {
  private static InputStream getResource(
    final String name)
    throws Exception {

    final String path = "/org/nypl/simplified/tests/opds/" + name;
    final URL url = OPDSFeedEntryParserTest.class.getResource(path);
    if (url == null) {
      throw new FileNotFoundException(path);
    }
    return url.openStream();
  }

  @Test
  public void testRoundTrip0()
    throws Exception {
    final OPDSAcquisitionFeedEntryParserType p =
      OPDSAcquisitionFeedEntryParser.newParser();

    final OPDSJSONParserType jp = OPDSJSONParser.newParser();

    final OPDSJSONSerializerType s = OPDSJSONSerializer.newSerializer();

    final InputStream rs0 =
      OPDSJSONSerializerTest.getResource("entry-0.xml");
    final OPDSAcquisitionFeedEntry e0 = p.parseEntryStream(URI.create("urn:test"), rs0);

    final ByteArrayOutputStream bao0 = new ByteArrayOutputStream();
    s.serializeToStream(s.serializeFeedEntry(e0), bao0);

    final InputStream rs1 = new ByteArrayInputStream(bao0.toByteArray());
    final OPDSAcquisitionFeedEntry e1 =
      jp.parseAcquisitionFeedEntryFromStream(rs1);

    {
      final List<OPDSAcquisition> e0a = e0.getAcquisitions();
      final List<OPDSAcquisition> e1a = e1.getAcquisitions();
      Assertions.assertEquals(e0a.size(), e1a.size());

      for (int index = 0; index < e0a.size(); ++index) {
        final OPDSAcquisition a0 = e0a.get(index);
        final OPDSAcquisition a1 = e1a.get(index);
        Assertions.assertEquals(a0.getRelation(), a1.getRelation());
        Assertions.assertEquals(a0.getType(), a1.getType());
        final Link u0 = a0.getUri();
        final Link u1 = a1.getUri();
        Assertions.assertEquals(u0.getBitrate(), u1.getBitrate());
        Assertions.assertEquals(u0.getDuration(), u1.getDuration());
        Assertions.assertEquals(u0.getHeight(), u1.getHeight());
        Assertions.assertEquals(u0.getHrefURI(), u1.getHrefURI());
        Assertions.assertEquals(u0.getRelation(), u1.getRelation());
        Assertions.assertEquals(u0.getTitle(), u1.getTitle());
        Assertions.assertEquals(u0.getType().toString(), u1.getType().toString());
        Assertions.assertEquals(u0.getWidth(), u1.getWidth());
        Assertions.assertEquals(a0.getIndirectAcquisitions(), a1.getIndirectAcquisitions());
      }

      Assertions.assertEquals(e0.getAvailability(), e1.getAvailability());
      Assertions.assertEquals(e0.getAuthors(), e1.getAuthors());
      Assertions.assertEquals(e0.getCategories(), e1.getCategories());
      Assertions.assertEquals(e0.getCover(), e1.getCover());
      Assertions.assertEquals(e0.getGroups(), e1.getGroups());
      Assertions.assertEquals(e0.getID(), e1.getID());
      Assertions.assertEquals(e0.getNarrators(), e1.getNarrators());
      Assertions.assertEquals(e0.getPreviewAcquisitions(), e1.getPreviewAcquisitions());
      Assertions.assertEquals(e0.getPublished(), e1.getPublished());
      Assertions.assertEquals(e0.getPublisher(), e1.getPublisher());
      Assertions.assertEquals(e0.getSummary(), e1.getSummary());
      Assertions.assertEquals(e0.getThumbnail(), e1.getThumbnail());
      Assertions.assertEquals(e0.getTitle(), e1.getTitle());
      Assertions.assertEquals(e0.getUpdated(), e1.getUpdated());
    }
  }

  @Test
  public void testRoundTrip1()
    throws Exception {
    final OPDSAcquisitionFeedEntryParserType ep =
      OPDSAcquisitionFeedEntryParser.newParser();

    final OPDSFeedParserType p = OPDSFeedParser.newParser(ep);
    final OPDSJSONParserType jp = OPDSJSONParser.newParser();

    final OPDSJSONSerializerType s = OPDSJSONSerializer.newSerializer();

    final InputStream rs0 = OPDSJSONSerializerTest.getResource("loans.xml");
    final OPDSAcquisitionFeed fe0 = p.parse(new URI("http://example.com"), rs0);

    final ByteArrayOutputStream bao0 = new ByteArrayOutputStream();
    s.serializeToStream(s.serializeFeed(fe0), bao0);

    final InputStream rs1 = new ByteArrayInputStream(bao0.toByteArray());
    final OPDSAcquisitionFeed fe1 = jp.parseAcquisitionFeedFromStream(rs1);

    {
      final List<OPDSAcquisitionFeedEntry> fe0e = fe0.getFeedEntries();
      final List<OPDSAcquisitionFeedEntry> fe1e = fe1.getFeedEntries();
      for (int index = 0; index < fe0e.size(); ++index) {
        final OPDSAcquisitionFeedEntry e0 = fe0e.get(index);
        final OPDSAcquisitionFeedEntry e1 = fe1e.get(index);
        Assertions.assertEquals(e0.getAcquisitions(), e1.getAcquisitions());
        Assertions.assertEquals(e0.getAuthors(), e1.getAuthors());
        Assertions.assertEquals(e0.getCategories(), e1.getCategories());
        Assertions.assertEquals(e0.getCover(), e1.getCover());
        Assertions.assertEquals(e0.getGroups(), e1.getGroups());
        Assertions.assertEquals(e0.getID(), e1.getID());
        Assertions.assertEquals(e0.getNarrators(), e1.getNarrators());
        Assertions.assertEquals(e0.getPreviewAcquisitions(), e1.getPreviewAcquisitions());
        Assertions.assertEquals(e0.getPublisher(), e1.getPublisher());
        Assertions.assertEquals(e0.getSummary(), e1.getSummary());
        Assertions.assertEquals(e0.getThumbnail(), e1.getThumbnail());
        Assertions.assertEquals(e0.getTitle(), e1.getTitle());
        // Forget comparing instances of Calendar, no implementation gets this
        // right
        // Assert.assertEquals(e0.getUpdated(), e1.getUpdated());
        // Assert.assertEquals(e0.getPublished(), e1.getPublished());
      }
    }
  }

  @Test
  public void testMimeTypes()
    throws Exception {
    final OPDSAcquisitionFeedEntryParserType p =
      OPDSAcquisitionFeedEntryParser.newParser();

    final OPDSJSONParserType jp = OPDSJSONParser.newParser();

    final OPDSJSONSerializerType s = OPDSJSONSerializer.newSerializer();

    final InputStream rs0 =
      OPDSJSONSerializerTest.getResource("entry-0.xml");
    final OPDSAcquisitionFeedEntry e0 = p.parseEntryStream(URI.create("urn:test"), rs0);

    final ByteArrayOutputStream bao0 = new ByteArrayOutputStream();
    s.serializeToStream(s.serializeFeedEntry(e0), bao0);

    final InputStream rs1 = new ByteArrayInputStream(bao0.toByteArray());
    final OPDSAcquisitionFeedEntry e1 =
      jp.parseAcquisitionFeedEntryFromStream(rs1);

    final List<OPDSAcquisition> e0a = e0.getAcquisitions();
    final List<OPDSAcquisition> e1a = e1.getAcquisitions();
    Assertions.assertEquals(e0a.size(), 1);
    Assertions.assertEquals(e0a.size(), e1a.size());

    final OPDSAcquisition e0FirstAcquisition = e0a.get(0);
    final OPDSAcquisition e1FirstAcquisition = e1a.get(0);

    Assertions.assertEquals(e0FirstAcquisition.getType().getSubtype(),
      e1FirstAcquisition.getType().getSubtype());
    Assertions.assertEquals(e0FirstAcquisition.getType().getType(),
      e1FirstAcquisition.getType().getType());
    Assertions.assertEquals(e0FirstAcquisition.getType().getFullType(),
      e1FirstAcquisition.getType().getFullType());
    Assertions.assertEquals(e0FirstAcquisition.getType().getParameters().size(),
      e1FirstAcquisition.getType().getParameters().size());

    for (String key : e0FirstAcquisition.getType().getParameters().keySet()) {
      Assertions.assertTrue(e1FirstAcquisition.getType().getParameters().containsKey(key));
      Assertions.assertEquals(e0FirstAcquisition.getType().getParameters().get(key),
        e1FirstAcquisition.getType().getParameters().get(key));
    }
  }
}

File: palace-tests/src/test/java/org/nypl/simplified/tests/opds/OPDSJSONSerializerTest.kt
package org.nypl.simplified.tests.opds

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntryParser
import org.nypl.simplified.opds.core.OPDSFeedParser.Companion.newParser
import org.nypl.simplified.opds.core.OPDSJSONParser
import org.nypl.simplified.opds.core.OPDSJSONSerializer.Companion.newSerializer
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.FileNotFoundException
import java.io.InputStream
import java.net.URI

class OPDSJSONSerializerTest {
  @Test
  @Throws(Exception::class)
  fun testRoundTrip0() {
    val p =
      OPDSAcquisitionFeedEntryParser.newParser()

    val jp = OPDSJSONParser.newParser()

    val s = newSerializer()

    val rs0: InputStream =
      getResource("entry-0.xml")
    val e0 = p.parseEntryStream(URI.create("urn:test"), rs0)

    val bao0 = ByteArrayOutputStream()
    s.serializeToStream(s.serializeFeedEntry(e0), bao0)

    val rs1: InputStream = ByteArrayInputStream(bao0.toByteArray())
    val e1 =
      jp.parseAcquisitionFeedEntryFromStream(rs1)

    run {
      val e0a = e0.acquisitions
      val e1a = e1.acquisitions
      Assertions.assertEquals(e0a.size, e1a.size)

      for (index in e0a.indices) {
        val a0 = e0a.get(index)
        val a1 = e1a.get(index)
        Assertions.assertEquals(a0.relation, a1.relation)
        Assertions.assertEquals(a0.type, a1.type)
        val u0 = a0.uri
        val u1 = a1.uri
        Assertions.assertEquals(u0.bitrate, u1.bitrate)
        Assertions.assertEquals(u0.duration, u1.duration)
        Assertions.assertEquals(u0.height, u1.height)
        Assertions.assertEquals(u0.hrefURI, u1.hrefURI)
        Assertions.assertEquals(u0.relation, u1.relation)
        Assertions.assertEquals(u0.title, u1.title)
        Assertions.assertEquals(u0.type.toString(), u1.type.toString())
        Assertions.assertEquals(u0.width, u1.width)
        Assertions.assertEquals(a0.indirectAcquisitions, a1.indirectAcquisitions)
      }

      Assertions.assertEquals(e0.availability, e1.availability)
      Assertions.assertEquals(e0.authors, e1.authors)
      Assertions.assertEquals(e0.categories, e1.categories)
      Assertions.assertEquals(e0.cover, e1.cover)
      Assertions.assertEquals(e0.groups, e1.groups)
      Assertions.assertEquals(e0.id, e1.id)
      Assertions.assertEquals(e0.narrators, e1.narrators)
      Assertions.assertEquals(e0.previewAcquisitions, e1.previewAcquisitions)
      // Assertions.assertEquals(e0.published, e1.published)
      Assertions.assertEquals(e0.publisher, e1.publisher)
      Assertions.assertEquals(e0.summary, e1.summary)
      Assertions.assertEquals(e0.thumbnail, e1.thumbnail)
      Assertions.assertEquals(e0.title, e1.title)
      Assertions.assertEquals(e0.updated, e1.updated)
    }
  }

  @Test
  @Throws(Exception::class)
  fun testRoundTrip1() {
    val ep =
      OPDSAcquisitionFeedEntryParser.newParser()

    val p = newParser(ep)
    val jp = OPDSJSONParser.newParser()

    val s = newSerializer()

    val rs0: InputStream = getResource("loans.xml")
    val fe0 = p.parse(URI("http://example.com"), rs0)

    val bao0 = ByteArrayOutputStream()
    s.serializeToStream(s.serializeFeed(fe0), bao0)

    val rs1: InputStream = ByteArrayInputStream(bao0.toByteArray())
    val fe1 = jp.parseAcquisitionFeedFromStream(rs1)

    run {
      val fe0e = fe0.feedEntries
      val fe1e = fe1.feedEntries
      for (index in fe0e.indices) {
        val e0 = fe0e.get(index)
        val e1 = fe1e.get(index)
        Assertions.assertEquals(e0.acquisitions, e1.acquisitions)
        Assertions.assertEquals(e0.authors, e1.authors)
        Assertions.assertEquals(e0.categories, e1.categories)
        Assertions.assertEquals(e0.cover, e1.cover)
        Assertions.assertEquals(e0.groups, e1.groups)
        Assertions.assertEquals(e0.id, e1.id)
        Assertions.assertEquals(e0.narrators, e1.narrators)
        Assertions.assertEquals(e0.previewAcquisitions, e1.previewAcquisitions)
        Assertions.assertEquals(e0.publisher, e1.publisher)
        Assertions.assertEquals(e0.summary, e1.summary)
        Assertions.assertEquals(e0.thumbnail, e1.thumbnail)
        Assertions.assertEquals(e0.title, e1.title)
        // Forget comparing instances of Calendar, no implementation gets this
        // right
        // Assert.assertEquals(e0.getUpdated(), e1.getUpdated());
        // Assert.assertEquals(e0.getPublished(), e1.getPublished());
      }
    }
  }

  @Test
  @Throws(Exception::class)
  fun testMimeTypes() {
    val p =
      OPDSAcquisitionFeedEntryParser.newParser()

    val jp = OPDSJSONParser.newParser()

    val s = newSerializer()

    val rs0: InputStream =
      getResource("entry-0.xml")
    val e0 = p.parseEntryStream(URI.create("urn:test"), rs0)

    val bao0 = ByteArrayOutputStream()
    s.serializeToStream(s.serializeFeedEntry(e0), bao0)

    val rs1: InputStream = ByteArrayInputStream(bao0.toByteArray())
    val e1 =
      jp.parseAcquisitionFeedEntryFromStream(rs1)

    val e0a = e0.acquisitions
    val e1a = e1.acquisitions
    Assertions.assertEquals(e0a.size, 1)
    Assertions.assertEquals(e0a.size, e1a.size)

    val e0FirstAcquisition = e0a.get(0)
    val e1FirstAcquisition = e1a.get(0)

    Assertions.assertEquals(
      e0FirstAcquisition.type.subtype,
      e1FirstAcquisition.type.subtype
    )
    Assertions.assertEquals(
      e0FirstAcquisition.type.type,
      e1FirstAcquisition.type.type
    )
    Assertions.assertEquals(
      e0FirstAcquisition.type.fullType,
      e1FirstAcquisition.type.fullType
    )
    Assertions.assertEquals(
      e0FirstAcquisition.type.parameters.size,
      e1FirstAcquisition.type.parameters.size
    )

    for (key in e0FirstAcquisition.type.parameters.keys) {
      Assertions.assertTrue(e1FirstAcquisition.type.parameters.containsKey(key))
      Assertions.assertEquals(
        e0FirstAcquisition.type.parameters.get(key),
        e1FirstAcquisition.type.parameters.get(key)
      )
    }
  }

  companion object {
    @Throws(Exception::class)
    private fun getResource(
      name: String
    ): InputStream {
      val path = "/org/nypl/simplified/tests/opds/" + name
      val url = OPDSFeedEntryParserTest::class.java.getResource(path)
      if (url == null) {
        throw FileNotFoundException(path)
      }
      return url.openStream()
    }
  }
}
